src/contract.rs

```diff
@@ -349,36 +367,56 @@ fn get_total_rewards_in_vault_denom(
     current_rewards
 }
 
-fn update_and_get_validators_to_record(
+fn get_validators_to_record(
     storage: &mut dyn Storage,
-    max_records_to_update_per_run: u32,
     timestamp: u64,
-) -> Result<Vec<ValidatorUpdateTimings>, ContractError> {
-    let mut validator_updates: Vec<ValidatorUpdateTimings> = vec![];
-    let mut records_to_update: u32 = 0;
+) -> Result<Vec<Addr>, ContractError> {
+    let batch_size = CONFIG.load(storage)?.batch_size;
+    let state = STATE.load(storage)?;
+    let last_cron_time_opt = state.cron_timestamps.last();
+    let validators = state.validators;
+    let total_validators: u32 = validators.len().try_into().unwrap();
+    let mut validator_index_for_next_cron = state.validator_index_for_next_cron;
+
+    // If the Cron time is completely New (Update State)
+    if last_cron_time_opt.is_none() || !last_cron_time_opt.unwrap().eq(&timestamp) {
+        STATE.update(storage, |mut s| -> StdResult<_> {
+            s.cron_timestamps.push(timestamp);
+            s.validator_index_for_next_cron = 0;
+            Ok(s)
+        })?;
+
+        validator_index_for_next_cron = 0
+    }
+
+    if validator_index_for_next_cron.ge(&total_validators) {
+        return Ok(vec![]);
+    }
+
+    let mut min = clamp(0, validator_index_for_next_cron, total_validators);
```

Comment:
Are you trying to return
`&validators[validator_index_for_next_cron..]` ?

Would you not be starting s.validator_index_for_next_cron with 0 during init message? So line 382 may not be required.

Reply:
Let's say we have 22 validators and our pagination is 10
Our **very first cron** We gave `t1` to this function from the cron

- run 1 => 0-10 | `validator_index_for_next_cron = 10` |  (line 382 = hey there are no timestamps so this must be the very first cron - Save the `t1` to timestamps history)
- run 2 => 10-20 | `validator_index_for_next_cron = 20` | (line 382 = hey its the same `t1` i last saved - Do nothing )
- run 3 => 20-22 | `validator_index_for_next_cron = 22` | (line 382 = same `t1` - Do nothing)
- run 4 => Nothing happens - early return  | (line 382 = same `t1` - Do nothing)
(Important: All the above runs got the same `t1` timestamp)

`validator_index_for_next_cron = 22` Now 


Now next day we gave the function timestamp `t2` from the cron
Now line 382 - Hey this is a new timestamp - reset things and save this new timestamp

Reply:
It's the same thing as checking 
```
if state.timestamps.last() == timestamp { reset param to 0 and insert into timestamps. }
```

My feedback was more around the clamp logic. You just need 
```
let start = validator_index_for_next_cron;
[&validators[(start+1)..(start+batch_size+1)]
```
This or without the +1's